lib/rust/data-structures/src/im_list.rs

```diff
@@ -0,0 +1,193 @@
+#[derive(Derivative, Deref, Debug)]
+#[derivative(Clone(bound = ""))]
+#[allow(missing_docs)]
+pub struct NonEmpty<T> {
+    pub node: Rc<Node<T>>,
+}
+
+/// A node of the [`List`]. Contains the current value and link to list [`tail`].
+#[derive(Clone, Debug)]
+#[allow(missing_docs)]
+pub struct Node<T> {
+    pub head: T,
+    pub tail: List<T>,
+}
+
+impl<T> Node<T> {
+    /// Constructor.
+    pub fn singleton(head: T) -> Self {
+        let tail = default();
+        Self { head, tail }
+    }
+}
+
+impl<T> NonEmpty<T> {
+    /// Constructor.
+    pub fn singleton(head: T) -> Self {
+        let node = Rc::new(Node::singleton(head));
+        Self { node }
+    }
+
+    /// Convert this non-empty list to list of unknown length.
+    pub fn into_list(self) -> List<T> {
+        let data = Some(self);
+        List { data }
+    }
+
+    /// Replace the head element of this list.
+    pub fn with_head(self, head: T) -> Self {
+        self.into_list().with_head(head)
+    }
+
+    /// Get the head element of this list.
+    pub fn head(&self) -> &T {
+        &self.head
+    }
+
+    /// Get tail of this list.
+    pub fn tail(&self) -> &List<T> {
+        &self.tail
+    }
+
+    /// Get the last element of this list.
+    pub fn last(&self) -> &T {
+        self.tail.last().unwrap_or(self.head())
+    }
+
+    /// Check whether this list is empty.
+    pub fn is_empty(&self) -> bool {
+        false
+    }
+
+    /// Convert this list to a vector.
+    fn to_vec(&self) -> Vec<&T> {
+        let mut out = vec![&self.head];
+        let mut list = self.tail();
+        loop {
+            match list.head() {
+                None => break,
+                Some(head) => {
+                    out.push(head);
+                    match list.tail() {
+                        None => break,
+                        Some(tail) => list = tail,
+                    }
+                }
+            }
+        }
+        out
+    }
+}
+
+impl<T> List<T> {
+    /// Replace head of this list.
+    pub fn with_head(self, head: T) -> NonEmpty<T> {
+        let tail = self;
+        let node = Rc::new(Node { head, tail });
+        NonEmpty { node }
+    }
+
+    /// Get the head element.
+    pub fn head(&self) -> Option<&T> {
+        self.as_ref().map(|t| t.head())
+    }
+
+    /// Get the tail of this list.
+    pub fn tail(&self) -> Option<&List<T>> {
+        self.as_ref().map(|t| t.tail())
+    }
+
+    /// Get the last element of this list.
+    pub fn last(&self) -> Option<&T> {
+        self.data.as_ref().map(|t| t.last())
+    }
+
+    /// Check whether this list is empty.
+    pub fn is_empty(&self) -> bool {
+        self.is_none()
+    }
+
+    /// Convert this list to a vector.
+    fn to_vec(&self) -> Vec<&T> {
+        self.data.as_ref().map(|t| t.to_vec()).unwrap_or_default()
+    }
+
+    /// Convert this list to a non-empty list. Return [`None`] if the list is empty.
+    pub fn try_to_non_empty(&self) -> &Option<NonEmpty<T>> {
```

Comment:
Yup, it should use the "as" word, but I'd not agree with `as_non_empty` because it is not `borrowed -> borrowed` function. It might not give the result, thus it should be named accordingly. The signature of `as_non_empty` should be `fn as_non_empty(&self) -> &NonEmpty<T>`. I'm changing that for now to `try_as_non_empty`, but I'd love to chat more about a better word than "try" here.